keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,626 @@
+- Cross-family connectivity: IPv4-to-IPv6 and IPv6-to-IPv4 connectivity is considered outside of the scope of this proposal. (As a possible future enhancement, the Kubernetes NGINX ingress controller could be modified to load balance to both IPv4 and IPv6 addresses for each endpoint. With such a change, it's possible that an external IPv4 client could access a Kubernetes service via an IPv6 pod address, and vice versa).
+- CNI network plugins: Some plugins other than the Bridge, PTP, and Host-Local IPAM plugins may support Kubernetes dual stack, but the development and testing of dual stack support for these other plugins is considered outside of the scope of this proposal.
+- Multiple IPs vs. Dual-Stack: Code changes will be done in a way to facilitate future expansion to more general multiple-IPs-per-pod and multiple-IPs-per-node support. However, this initial release will impose "dual-stack-centric" IP address limits as follows:
+  - Pod addresses: 1 IPv4 address and 1 IPv6 addresses per pod maximum
+  - Node addresses: 1 IPv4 address and 1 IPv6 addresses per pod maximum
+  - Service addresses: 1 service IP address per service
+- Kube-DNS is expected to be End-of-Life soon, so dual-stack testing will be performed using coreDNS.
+- External load balancers that rely on Kubernetes services for load balancing functionality will only work with the IP family that matches the IP family of the cluster's service CIDR.
+- Dual-stack support for Kubernetes orchestration tools other than kubeadm (e.g. miniKube, KubeSpray, etc.) are considered outside of the scope of this proposal.
+
+## Proposal
+
+In order to support dual-stack in Kubernetes clusters, Kubernetes needs to have awareness of and support dual-stack addresses for pods and nodes. Here is a summary of the proposal (details follow in subsequent sections):
+
+- Kubernetes needs to be made aware of multiple IPs per pod (limited to one IPv4 and one IPv6 address per pod maximum).
+- Link Local Addresses (LLAs) on a pod will remain implicit (Kubernetes will not display nor track these addresses).
+- For simplicity, only a single family of service IPs per cluster will be supported (i.e. service IPs are either all IPv4 or all IPv6).
+- Backend pods for a service can be dual stack.
+- Endpoints for a dual-stack backend pod will be represented as a dual-stack address pair (i.e. 1 IPv4/IPv6 endpoint per backend pod, rather than 2 single-family endpoints per backend pod)
```

Comment:
I think the non-goal is correct -- Kubernetes itself is NOT doing address family translation.  The fact that Ingress controllers can do that is merely a side-effect of the fact that they are almost universally full proxies which accept the frontside connection and open another for the backside.  I don't think it is obvious that we won't convert a v4 to a v6 connection via service VIP, and we should be clear that it is NOT part of the scope.

When I wrote this question I was thinking about Service VIP -> Backends.  That has to be same-family (because of the non-goal above), so alt-family endpoints is nonsensical.  I think this is also true for external IPs and load-balancer IPs -- they are received and processed the same as cluster IPs, with no family conversion.

BUT...

1) external IPs is a list, so could plausibly include both families.
2) LB status.ingress is a list, so could plausibly include both families.
3) nodePorts could reasonably be expected to work on any interface address.
4) headless services could reasonably be expected to work on any interface address.

You suggest you might be willing to step back from (4), but unless we also step back from (1), (2), AND (3), that wouldn't save any work.

I think the only simplification here is if we can say "all service-related logic is single family".  And I am not sure that is very useful -- tell me if you disagree.

Assuming we have ANY service-related functionality with both families, we need dualstack endpoints :(

Or am I missing something?

Reply:
...more to the point, is multi-family endpoints, nodeports, LBs, etc somethign we can defer to a "phase 2" and iterate?  Would it simplify this proposal or just make it not useful?

Reply:
OK, so any scope-reduction around not doing dual-stack endpoints is rendered moot,a nd all such comments should be ignored :)